### DISCUSSION TITLE
HTTP 메서드 중 PATCH와 PUT의 차이점을 말해주세요.

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#147] [Network]
뭘까용

**Comment by jimi567** [comment]
HTTP 메서드 중 PATCH와 PUT은 모두 리소스를 업데이트 하는 목적으로 사용된다. 하지만 두 개의 메서드에는 다음과 같은 명확한 차이점이 있다.

> 1. PUT

* 목적 : 클라이언트가 서버에게 리소스의 특정 위치에 대해 **완전한 업데이트** 를 요청하는 것
* 클라이언트는 요청 바디에 업데이트할 리소스의 전체 Content를 포함시켜야함(만약 필드를 수락한다면 해당 필드는 null로 업데이트 됨)
* 서버는 해당 위치에 요청된 리소스를 업데이트 하거나 새로 생성
* 멱등성 관점에서는 항상 **멱등함**

> 2. PATCH 

* 목적 : 클라이언트가 서버에게 특정 리소스의 **일부를 업데이트** 할 때 사용된다.
* 클라이언트는 요청 바디에 실제로 변경하려는 리소스의 일부 필드만 포함시킨다.
* 멱등성 관점에서는 항상 **멱등하다고 볼 수 없음**

간단하게 얘기하자면, PUT은 전체를 업데이트. PATCH는 일부만 업데이트
PUT은 멱등하지만, PATCH는 멱등하지 않음.

---
헷갈릴 수 있어서, 예시도 추가했습니다.

아래와 같은 리소스가 존재한다고 가정.

|id|name|age|
|--|--|--|
|1|위브|21|
|2|아톰|22|
|3|프린|23|

---- 
> 일부 필드만 업데이트 하는 경우

1. 위브의 나이를 22살로 PUT
PUT /crew?id=1
{  
   age : 22
}

결과
|id|name|age|
|--|--|--|
|1|null|22|
|2|아톰|22|
|3|프린|23|


2. 위브의 나이를 22살로 PATCH
PATCH /crew?id=1
{  
   age : 22
}

결과
|id|name|age|
|--|--|--|
|1|위브|22|
|2|아톰|22|
|3|프린|23|

-- > PUT은 보내지 않은 필드값은 null로 대체됨

----

> 요청한 URI에 실제 자원이 존재하지 않은 경우


1. 4번 ID에 토미 30살로 PUT
PUT /crew?id=4
{  
   name : 토미
   age : 30
}

결과
|id|name|age|
|--|--|--|
|1|위브|21|
|2|아톰|22|
|3|프린|23|
|4|토미|30|


2. 4번 ID에 토미 30살로 PATCH
PATCH /crew?id=4
{  
   name : 토미
   age : 30
}

결과
일반적으로 404를 응답하는 등 오류를 응답

---
만약 서버에 기존 나이를 1살 증가시키는 API가 존재한다면, PATCH는 멱등성을 보장하지 못함

PATCH /crew?id=1
{
   $increase: 'age',
    value: 1
}

이런 요청을 여러 번 보내게 된다면, 리소스의 상태는 매번 바뀌게 됨.

**Comment by GIVEN53** [comment]
PATCH와 PUT은 모두 HTTP 프로토콜에서 리소스를 업데이트하는 메서드입니다.

### 차이점
#### 1. 관점
PUT: 리소스를 **대체(Replace)** 하는 목적을 가집니다.
클라이언트가 PUT을 사용하여 특정 리소스의 전체 데이터를 업데이트하고, 이때 클라이언트가 제공하는 데이터로 리소스의 전체 내용이 교체됩니다. 즉, 클라이언트가 전체 리소스를 알고 있는 상황에서 사용됩니다.
보내지 않은 값은 **null로 대체**됩니다.

PATCH: 리소스의 일부를 **수정(Update)** 하는 목적을 가집니다.
클라이언트가 PATCH를 사용하여 리소스의 일부 데이터만을 업데이트하고, 이때 클라이언트가 제공하는 데이터로 리소스의 일부 내용이 변경됩니다. PATCH는 리소스의 일부분만 알고 있는 상황에서 사용됩니다.
보내지 않은 값은 **그대로 유지**됩니다.

#### 2. 요청 URI에 자원이 존재하지 않을 때
PUT: 새로운 자원을 생성합니다.

PATCH: 새로운 자원을 생성하지 않고, 서버는 오류를 응답합니다.

#### 3. 안전성
PUT: 리소스의 전체 데이터가 변경되므로 실수로 인해 리소스의 손상이 발생할 수 있습니다. 예를 들어, 클라이언트가 리소스를 완전히 잘못된 데이터로 대체할 수 있습니다.

PATCH: 일부 데이터만 변경하기 때문에, 리소스의 전체 데이터가 잘못되는 것을 방지할 수 있습니다. 클라이언트가 변경하려는 데이터의 일부분만 알고 있을 때 유용하게 사용됩니다.

#### 4. 멱등성(Idempotent)
PUT: 멱등성 특성을 가집니다.
PATCH: 구현에 따라 멱등성이 보장될 수도 있고 보장되지 않을 수도 있습니다.

### 결론
PUT과 PATCH는 모두 리소스 업데이트를 목적으로 하지만, PUT은 전체 리소스를 대체하는 데 사용되고, PATCH는 일부 데이터만을 수정하는 데 사용됩니다. PUT은 리소스의 전체 정보를 갖고 있는 경우 사용하고, PATCH는 리소스의 일부 정보만을 변경할 때 사용하는 것이 적합합니다.

**Comment by 3Juhwan** [comment]
PATCH가 멱등성이 보장되지 않을 수 있는 경우는 다음과 같습니다. 

### 멱등성 보장 O
- `{"age": 10}` 이라는 자원이 존재합니다. 
- 나이를 11살로 갱신합니다. `PATCH {"data": 1 }`
- `{"age": 11}` 로 변경됩니다. 

### 멱등성 보장 X
- `{"age": 10}` 이라는 자원이 존재합니다. 
- 나이를 1살 더합니다. `PATCH {"data": 1 }`
- `{"age": 11}` 로 변경됩니다. 
- 나이를 1살 더합니다. `PATCH {"data": 1 }`
- `{"age": 12}` 로 변경됩니다. 

출처: https://www.inflearn.com/questions/110644/patch-%EB%A9%94%EC%84%9C%EB%93%9C%EA%B0%80-%EB%A9%B1%EB%93%B1%EC%9D%B4-%EC%95%84%EB%8B%8C-%EC%9D%B4%EC%9C%A0
